### ISSUE TITLE
Scaling SA to 150+ addons: performance and memory usage

### ISSUE THREAD

**Issue by WorldLanguages** [ScratchAddons/ScratchAddons#5190]
We're getting close to the 150+ addon count.
This issue intends to list the technical side of handling this amount of addons. The user experience (UX) side of having so many addons belongs in a different issue.

### Memory usage (background page)
It's important to consider the memory usage of our background page. It's true that Scratch itself is pretty memory intensive, but SA is running all the time, even if the user isn't currently using Scratch!
Since v1.0.0, all manifests are stored inside a variable in the background page.
This includes all this data from all addons, no matter if they're enabled or not:
  - Addon ID
  - The entire addon.json file as an Object
  - If the browser language is not set to English, and it's a language we support (such as Spanish), the English name and description are stored in memory too (see https://github.com/ScratchAddons/ScratchAddons/blob/master/background/load-addon-manifests.js#L58)

This `scratchAddons.manifests` array gets larger with every release. I'm not sure how to properly check the actual memory usage of this "big" object, but according to Chrome Devtools, it appears to use about 0.25MB.

We could also check whether `scratchAddons.l10n` stores any unnecessary messages in memory in the background page.

### chrome.storage.sync per-key size limit

Related: #4721

As we add more settings each release, we'll soon reach 8192 bytes for the `addonSettings` key.
See QUOTA_BYTES_PER_ITEM on https://developer.chrome.com/docs/extensions/reference/storage/
It was previously easy to reach this limit by exporting/importing settings, but we fixed it. However, the problem will return at some point, if we keep adding new settings.

### Settings page performance

This is less important than the other things I mentioned above, but it belongs on this list.
We're using Vue v1 and I've implemented some weird performance hacks to try to avoid performance bottlenecks.
However, of course the settings page has more DOM nodes each release, as well as more JavaScript (vdom?) objects handled by Vue.
However, the solution here is more straight forward, we could just migrate to Vue 3, and we'll get better performance with less hacks. We would also avoid shipping the CSP version which is slower.

### Anything else we need to consider?

Please tell me whether I'm missing something here.

**Comment by iqnite** [comment]
Performance isn't the only problem: just think about the SSD storage it will need when it will have 250 addons with all the GUI icons that belong to them. Actually, it's not relevant, but like in the boiled frogs principle, Scratch Addons will grow and grow until it will be too big.

**Comment by iqnite** [comment]
Anyways, a possible solution for the background performance usage would be that Scratch Addons simply detects if there aren't any scratch.mit.edu or Scratch Addons internal pages opened and stop all processes that are not critical. I think that would already make a big difference.

**Comment by DNin01** [comment]
> Performance isn't the only problem: just think about the SSD storage it will need when it has 250 addons with all the GUI icons that belong to them. It's not relevant right now, but like in the boiled frogs principle, Scratch Addons will grow and grow until it is too big.

I don't think Scratch Addons will take up too much space. My installation's size is currently 5.9 MB, which is about the same as a picture or two. It might be worse if the addons get thumbnails, though (unless those are loaded from the Scratch Addons website).

**Comment by WorldLanguages** [comment]
> unless those are loaded from the Scratch Addons website

That will probably be the case.

**Comment by DNin01** [comment]
I think we could potentially use the `addonSettings` key for general addon settings, but addons with more settings could have their own settings keys so SA doesn't hit the QUOTA_BYTES_PER_ITEM limit, especially for `discuss-button` and `cloud-games`, which have [table settings](https://github.com/ScratchAddons/ScratchAddons/blob/b32ce769838f37c8d2da391fc32b8777dad1f18d/addons/discuss-button/addon.json#L34) that could cause this limit to be hit real fast.

**Comment by WorldLanguages** [comment]
@DNin01 I think it would be simpler to have `addonSettings1`, `addonSettings2` and `addonSettings3`. Realistically, we won't be filling 3 of these any time soon.
We could separate those 3 by addon ID, such as those that start with a-e, then f-o and lastly p-z.

**Comment by mxmou** [comment]
> @DNin01 I think it would be simpler to have `addonSettings1`, `addonSettings2` and `addonSettings3`. Realistically, we won't be filling 3 of these any time soon. We could separate those 3 by addon ID, such as those that start with a-e, then f-o and lastly p-z.

I don't think that's simpler than having one key for each addon.

**Comment by WorldLanguages** [comment]
@mxmou I don't think it would be very different in terms of complexity. `if (addonId === "cloud-games")` would change to `if (addonId[0].match(/[a-e]/))`.
Having 3 different storage keys would also be a permanent solution that would stay untouched for years, even if there's more table settings in the future.

**Comment by WorldLanguages** [comment]
> I think it would be simpler to have `addonSettings1`, `addonSettings2` and `addonSettings3`. Realistically, we won't be filling 3 of these any time soon. We could separate those 3 by addon ID, such as those that start with a-e, then f-o and lastly p-z.

@apple502j What do you think about this idea?

**Comment by Hans5958** [comment]
Friendly reminder: As of v1.30.0, the addon count is now 151. In such offtopic sense, I have edited the parent post, but something needs be done.

**Comment by WorldLanguages** [comment]
Looks like I should have used the "awaiting response" label above :P The bot would have responded after 7 days saying the issue is stale

**Comment by DNin01** [comment]
The `cloud-games` addon has [table settings](https://github.com/ScratchAddons/ScratchAddons/blob/master/addons/cloud-games/addon.json#L27) with `https://scratch.mit.edu/` at the start of every single item. That could add up. Considering that `discuss-button` doesn't, we could probably remove that part from `cloud-games` - it's essentially a redundant prefix.

**Comment by WorldLanguages** [comment]
@DNin01 I don't think that's a big deal.

**Comment by DNin01** [comment]
> @DNin01 I don't think that's a big deal.

I could still imagine someone overflowing the storage by adding upwards of ten projects to the list. I might be that kind of person if I used that addon often. Maybe that won't happen enough to be an issue, though.

**Comment by DNin01** [comment]
> @DNin01 I think it would be simpler to have `addonSettings1`, `addonSettings2` and `addonSettings3`. Realistically, we won't be filling 3 of these any time soon. We could separate those 3 by addon ID, such as those that start with a-e, then f-o and lastly p-z.

Maybe we could have separate settings files for the addons that have a table in their settings or have a lot of settings (like customizable colors addons)?

**Comment by DNin01** [comment]
> @DNin01 I don't think it's necessary to have a single key for each addon with table settings.

We could also do one file for every table setting.

**Comment by WorldLanguages** [comment]
@DNin01 I think you're referring to "storage keys" by using the word "file".

**Comment by WorldLanguages** [comment]
Adding tag to get a self-reminder in 1 week: we should create a dedicated issue about partitioning the `addonSettings` key.

**Comment by WorldLanguages** [comment]
> Adding tag to get a self-reminder in 1 week: we should create a dedicated issue about partitioning the `addonSettings` key.

Created issue: #5849

**Comment by DNin01** [comment]
We can probably cross off the addons memory usage issue since that was concluded to be not worth resolving.

**Comment by Samq64** [comment]
Removing all the empty `settings` objects will reduce the `chrome.sync` storage used by about 1.3KB on a fresh install.
